feat!: OpenTelemetry metrics (#13232)
argoproj/argo-workflows: docs/metrics.md

Emitting custom metrics with Argo is easy, but it's important to understand what makes a good Prometheus metric and the
best way to define metrics in Argo to avoid problems such as [cardinality explosion](https://stackoverflow.com/questions/46373442/how-dangerous-are-high-cardinality-labels-in-prometheus).

Metrics can be collected using the OpenTelemetry protocol or via prometheus compatible scraping.

## Metrics configuration

It is possible to collect metrics via the OpenTelemetry protocol or via prometheus compatible scraping. Both of these mechanisms can be enabled at the same time, which could be useful if you'd like to migrate from one system to the other. Using multiple protocols at the same time is not intended for long term use however.

OpenTelemetry is the recommended way of collecting metrics, and the OpenTelemetry collector can export metrics to prometheus via [the prometheus remote write exporter](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/exporter/prometheusremotewriteexporter/README.md).

### OpenTelemetry protocol

> To enable the OpenTelemetry protocol you must set the environment variable `OTEL_EXPORTER_OTLP_ENDPOINT` or `OTEL_EXPORTER_OTLP_METRICS_ENDPOINT`. It will not default to the default value if left blank, instead no Open

Reviewer: ```suggestion
To enable the OpenTelemetry protocol you must set the environment variable `OTEL_EXPORTER_OTLP_ENDPOINT` or `OTEL_EXPORTER_OTLP_METRICS_ENDPOINT`.
It will not use the default value if left blank, instead no Open
```

- simplify, per k8s style guide
- prefer 1 sentence per line, per style guide
- incomplete sentence?

  port: 8080

  # IgnoreErrors is a flag that instructs prometheus to ignore metric emission errors. Default is "false"
  ignoreErrors: false

  # Use a self-signed cert for TLS, default true
  secure: true
```

The metric names emitted by this mechanism are prefixed with `argo_workflows_`. `Attributes` are exposed as prometheus `labels` of the same name.

### Common

You can adjust various elements of the metrics configuration by changing values in the [Workflow Controller Config Map](workflow-controller-configmap.md).

```yaml
metricsConfig: |
  # MetricsTTL sets how often custom metrics are cleared from memory. Default is "0", metrics are never cleared
  metricsTTL: "10m"
  options:
    pod_missing:
      disable: true
    cronworkflows_triggered_total:
      disabledAttributes:
        - name
    k8s_request_duration:
      histogramBuckets: [ 1.0, 2.0, 10.0 ]
```

### Options

Using options you can manipulate the metrics created by the workflow controller. These options apply to the built-in metrics and any custom metrics you create. Each option applies to the named metric only, and applies to all output methods.

> `disable: true` will disable the emission of the metric from the system.

Reviewer: Typically the convention in the ecosystem, such as Helm, is `enabled: false` -- note that this includes `metricsConfig.enabled` just above this.

The tense and convention is important to keep consistent and predictable IMO

Author: Yeah, this and the 'options' naming were just where I'd ended up and wanted to see if my changes passed in proper CI.

Metrics that inform on the state of a Workflow, or a series of Workflows. These custom metrics are defined by the user in the Workflow spec.

Emitting custom metrics is the responsibility of the emitter owner. Since the user defines Workflows in Argo, the user is responsible for emitting metrics correctly.

> Currently custom metrics and their labels must be valid prometheus and OpenTelemetry metric names, which limits them to alpha-numerics, `_`. This applies even if you're only using OpenTelemetry for metrics.

Reviewer: ```suggestion
Currently custom metrics and their labels must be valid Prometheus and OpenTelemetry metric names, which limits them to alphanumeric characters and `_`.
This applies even if you're only using OpenTelemetry for metrics.
```

- use consistent wording for "alphanumeric characters"
- fix missing "and"
- prefer 1 sentence per line of markdown, per style guide

This can also likely be simplified, per k8s style guide

Metrics are also ephemeral, meaning there is no guarantee that they will be persisted for any amount of time. If you need
a way to view and analyze historical data, consider the [workflow archive](workflow-archive.md) or reporting to logs.

### Counter, gauge and histogram

> The following terms have specific meanings for the metrics that we collect:
> 
> - `counter`: A value that accumulates over time – you can think of this like an odometer on a car; it only ever goes up.
>     - The rate of change of counters can be a very powerful tool in understanding these metrics.
> - `gauge`: Measures a current value at the time it is read. An example would be the fuel gauge in a vehicle.
> - `histogram`: A client-side aggregation of values, such as request latencies. A histogram is a good choice if you are interested in value statistics. For example: How many requests take fewer than 1 second?
> 
> Thanks to [OpenTelemetry](https://opentelemetry.io/docs/concepts/signals/metrics/#metric-instruments) for these explanations.

Reviewer: ```suggestion
These terms are [defined by OpenTelemetry](https://opentelemetry.io/docs/concepts/signals/metrics/#metric-instruments):

> - `counter`: A value that accumulates over time – you can think of this like an odometer on a car; it only ever goes up.
    - The rate of change of counters can be a very powerful tool in understanding these metrics.
> - `gauge`: Measures a current value at the time it is read. An example would be the fuel gauge in a vehicle.
> - `histogram`: A client-side aggregation of values, such as request latencies. A histogram is a good choice if you are interested in value statistics. For example: How many requests take fewer than 1 second?

```

This looks to be almost entirely a direct quote, so we should quote it. Though I would prefer to just link to it if we have no notable differences

Author: We have no notable differences, but I felt it was worth including directly as it's obviously been an area of confusion and I want to ensure users fully grasp the concepts.

### Default Controller Metrics

> Metrics for the Four Golden Signals are:

Reviewer: ```suggestion
Metrics for the [Four Golden Signals](https://sre.google/sre-book/monitoring-distributed-systems/#xref_monitoring_golden-signals) are:
```

let's add a link while we're here

- Latency: `queue_latency`
- Traffic: `gauge` and `queue_depth_gauge`
- Errors: `count` and `error_count`
- Saturation: `workers_busy` and `workflow_condition`

!!! High cardinality

>     Some metric attributes may have high cardinality and are marked with ⚠️ to warn you. You may need to disable this metric or disable the attribute.

Reviewer: ```suggestion
    Some metric attributes may have high cardinality and are marked with ⚠️.
    You may want to disable the metric or attribute.
    
```

- We should use proper [`mkdocs-material` icons](https://squidfunk.github.io/mkdocs-material/reference/icons-emojis/) for this icon
- we should explain _why_ they may want to disable it, such as high storage usage or similar. Idk if there's a link we can provide for that, but that would be ideal
- Missing new line after note
- prefer 1 sentence per line of markdown, per [style guide](https://argo-workflows.readthedocs.io/en/release-3.5/doc-changes/)

Author: I don't see an precedent for usage of non-emoji based icons. I prefer the emoji based ones because all text editors render them as icons.
An explanation of high cardinality is mentioned at least twice elsewhere including a link to an explanation. I felt it was redundant to explain this more.

#### `build_info`

The build information for this workflow controller

| attribute   | explanation                                                            |
|-------------|------------------------------------------------------------------------|
| `version`   | The version of argo-workflows                                          |
| `platform`  | Platform this is running on, as go describes it e.g. `linux/amd64` |
| `gover`     | Version of go that built this workflow controller                  |
| `build`     | Build date for this workflow controller                                |
| `compiler`  | The compiler used to build this workflow controller e.g. `gc`          |
| `commit`    | The full git SHA1 of the code that built this workflow controller      |
| `treestate` | Whether the git tree was `dirty` or `clean` when this was built        |
| `tag`       | The tag on the git commit or `untagged` if it was not tagged           |

#### `cronworkflows_triggered_total`

A counter of the number of times a CronWorkflow has been

| attribute   | explanation                               |
|-------------|-------------------------------------------|

> | `name`     | ⚠️ The name of the CronWorkflow. |

Reviewer: I think it would make more sense to have the icon next to the attribute rather than its explanation -- much easier to see that way:

```suggestion
| ⚠️ `name`     | The name of the CronWorkflow. |
```

Also we should consider automatically generating all these from the go types as we do with the Fields Reference and others

Author: Automatically generating would be lovely, but more scope creep, so not doing it now.

Author: I disagree on the icon positioning, and prefer it with the description. It is in effect a concise part of the description, and I don't feel it is important enough to warrant moving left.

Reviewer: Since you did write a "High Cardinality" warning already, it did sound important enough 😅

[Argo variables](variables.md) can be included anywhere in the metric spec, such as in `labels`, `name`, `help`, `when`, etc.

> Metric names can only contain alpha-numeric characters, `_`, and `:`.

Reviewer: ```suggestion
Metric names can only contain alphanumeric characters, `_`, and `:`.
```

"[alphanumeric](https://www.merriam-webster.com/dictionary/alphanumeric)" is a word with no hyphen
